pkg/controller/component/vars.go

```diff
@@ -373,6 +379,75 @@ func buildEnv4UserDefined(annotations map[string]string) ([]corev1.EnvVar, error
 	return vars, nil
 }
 
+func evaluateObjectVarsExpression(definedVars []appsv1alpha1.EnvVar, credentialVars []corev1.EnvVar, vars *[]corev1.EnvVar) error {
+	var (
+		isValues = make(map[string]bool)
+		values   = make(map[string]string)
+	)
+	normalize := func(name string) string {
+		return strings.ReplaceAll(name, "-", "_")
+	}
+	for _, v := range [][]corev1.EnvVar{*vars, credentialVars} {
+		for _, vv := range v {
+			if vv.ValueFrom == nil {
+				isValues[vv.Name] = true
+				values[normalize(vv.Name)] = vv.Value
+			} else {
+				isValues[vv.Name] = false
+			}
+		}
+	}
+
+	evaluable := func(v appsv1alpha1.EnvVar) bool {
+		if v.Expression == nil || len(*v.Expression) == 0 {
+			return false
+		}
+		isValue, ok := isValues[v.Name]
+		// !ok is for vars that defined and resolved successfully, but have nil value.
+		return !ok || isValue
+	}
+
+	update := func(name, value string) {
+		if val, exist := values[normalize(name)]; exist {
+			if val != value {
+				for i := range *vars {
+					if (*vars)[i].Name == name {
+						(*vars)[i].Value = value
+						break
+					}
+				}
+			}
+		} else {
+			// TODO: insert the var to keep orders?
+			*vars = append(*vars, corev1.EnvVar{Name: name, Value: value})
+		}
+		values[normalize(name)] = value
+	}
+
+	eval := func(v appsv1alpha1.EnvVar) error {
+		if !evaluable(v) {
+			return nil
+		}
+		tpl, err := varTemplate.Parse(*v.Expression)
+		if err != nil {
+			return err
+		}
+		var buf strings.Builder
+		if err = tpl.Execute(&buf, values); err != nil {
```

Comment:
So this is the reason why the order of variable definitions containing expressions is significant. Variables defined later in the sequence can reference the values of variables defined earlier.